Solutions/Marain.UserNotifications.Client/Marain/UserNotifications/Client/ApiDeliveryChannel/Resources/NotificationResource.cs

```diff
@@ -0,0 +1,237 @@
+﻿// <copyright file="NotificationResource.cs" company="Endjin Limited">
+// Copyright (c) Endjin Limited. All rights reserved.
+// </copyright>
+
+namespace Marain.UserNotifications.Client.ApiDeliveryChannel.Resources
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Text.Json;
+    using System.Text.Json.Serialization;
+    using Menes.Hal;
+
+    /// <summary>
+    /// A notification.
+    /// </summary>
+    [JsonConverter(typeof(NotificationResourceConverter))]
+    public readonly struct NotificationResource : IHalDocumentResource
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="NotificationResource"/> class.
+        /// </summary>
+        /// <param name="source">The underlying JSON document for the object.</param>
+        public NotificationResource(in HalDocumentResource source)
+        {
+            if (!source.IsNotificationResource())
+            {
+                throw new ArgumentException("The supplied document is not a NotificationResource", nameof(source));
+            }
+
+            this.HalDocument = source;
+        }
+
+        /// <inheritdoc/>
+        public HalDocumentResource HalDocument { get; }
```

Comment:
As mentioned elsewhere, the System.Text.Json/aspnetcore handler path, if we are careful, will let us avoid most of the allocations for this stuff that just wraps the JsonElements read from / written to the stream. If you remember from a few weeks ago I benchmarked a demo of this that demonstrate a seriously reduced amount of alllocation v. the Json.NET / serialization approach. That's why these are modelled like this.

Reply:
Ah, but I don't think that the savings you demonstrated have much to do with the use of `struct`. Jon sent me a copy of your benchmark, and I've modified it by adding a parallel hierarchy of types—we now have `HalDocumentResourceClass`, `IHalDocumentResourceClass`, `PersonResourceClass`, `PersonResourceClassExtensions`, `PersonNameResourceClass`, `PersonNameResourceClassExtensions`, and `PersonNameResourceClassUserExtensions`, in which all I've replaced `struct` with `class`. So I can compare perf with that and the struct version:

```
|                                         Method |      Mean |     Error |    StdDev |       Min |       Max |    Median |  Gen 0 |  Gen 1 | Gen 2 | Allocated |
|----------------------------------------------- |----------:|----------:|----------:|----------:|----------:|----------:|-------:|-------:|------:|----------:|
| UsingStronglyTypedHalDocumentsAndSerialization |  5.867 us | 1.0720 us | 0.7091 us |  4.886 us |  6.627 us |  6.095 us | 0.2441 |      - |     - |   2.02 KB |
|        UsingValueTypeStronglyTypedHalDocuments |  3.572 us | 0.0414 us | 0.0274 us |  3.531 us |  3.620 us |  3.569 us | 0.2594 |      - |     - |   2.08 KB |
|         UsingRefTypedStronglyTypedHalDocuments |  3.575 us | 0.2779 us | 0.1654 us |  3.470 us |  3.954 us |  3.498 us | 0.2365 |      - |     - |   1.93 KB |
|                           UsingRawHalDocuments |  3.422 us | 0.0343 us | 0.0204 us |  3.387 us |  3.442 us |  3.432 us | 0.2289 |      - |     - |   1.84 KB |
|                               UsingJsonNetLinq |  7.317 us | 0.4107 us | 0.2716 us |  7.032 us |  7.790 us |  7.241 us | 1.2817 | 0.0458 |     - |   10.5 KB |
|                      UsingJsonNetSerialization | 14.568 us | 0.1563 us | 0.0930 us | 14.447 us | 14.701 us | 14.588 us | 1.6937 | 0.0305 |     - |  13.91 KB |
```

This seems to show that the bulk of the perf and memory wins come from the "read out of the hal document in place" approach vs full deserialization. The further wins from using `struct` instead of `class` seem very small by comparison, to the point that I'm not convinced it's worth the pain. If you care that much about these additional marginal wins then maybe you should go with the `UsingRawHalDocuments` approach, because that's even better.

And for serialization, mysteriously we seem to get lower memory usage with classes:

```
|                             Method |     Mean |     Error |    StdDev |      Min |      Max |   Median |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------------------- |---------:|----------:|----------:|---------:|---------:|---------:|-------:|------:|------:|----------:|
| UsingHalDocumentValueSerialization | 1.763 us | 0.0217 us | 0.0144 us | 1.742 us | 1.783 us | 1.763 us | 0.1087 |     - |     - |     920 B |
| UsingHalDocumentClassSerialization | 1.729 us | 0.0336 us | 0.0176 us | 1.705 us | 1.756 us | 1.726 us | 0.1011 |     - |     - |     856 B |
|          UsingJsonNetSerialization | 2.858 us | 0.1052 us | 0.0550 us | 2.765 us | 2.923 us | 2.864 us | 0.3433 |     - |     - |    2896 B |
```

Reply:
Ok! Well, that's telling. I remain baffled as to how that contrives to do less allocation.

Immutable classes it is then.

Reply:
It may be related to the fact that all these structs implement interfaces. That makes it very easy to box by accident—any time you try to use one of these things through an interface it'll generate a new box for it, whereas with a class, you get exactly as many as you `new` up, no more, no less, no matter how many times you convert it to the interface type.

Reply:
Could be - although when this approach has been used elsewhere we've been fastidious to use the generic types with the interface constraint to avoid boxing.

Reply:
I was a bit premature with "Immutable classes it is". There's something deeper we need to understand to ensure this endz up being the skinniest possible wrapper, and that the usage patterns are clarified.

Reply:
https://github.com/marain-dotnet/Marain.UserNotifications/issues/35